chef/automate
# Move a2ha's automate backend deployment to automate repo

### :nut_and_bolt: Description: What code changed, and why?

For Code Movement to the Automate Repo we have 12 components, so as part of this PR we are targeting below components.
automate-backend-deployment

- As part of this PR. Below things is added in automate repo.

   **terraform**
   - A2ha's terraform code is added in automate's terraform dir. name of a2ha's terraform module is a2ha-terraform.
   inspec(below a2ha's inspec is added in automate's inspec dir)
   - automate-backend-elasticsearch-smoke
   - automate-backend-postgresql-smoke
   - automate-backend-resources
   - automate-frontend-chef-server-smoke
   - automate-frontend-smoke
  
   **test**
   - Before a2ha deployment and mid deployment test is written in test dir. automate doesn't have any test dir so this is new dir.
   
   **scripts**
   - airgap_bundle.sh
   - bumpaib.sh
   - bundle_creation.sh
   - credentials
   - gather-logs
   - generate-chef-load-user-toml.sh
   - smoke-test
   - sup-keys.sh
   - tf_var_sort.py
 
   **Makefile**
   - a2ha's makefile is added in a2ha-terraform dir
   
    **Certs**
    - This cert will be needed to rotate a2ha certificate

### :chains: Related Resources
https://app.zenhub.com/workspaces/the-delta-quadrant-5fc7c4c3922dc10021323fa6/issues/chef/automate/5443
### :+1: Definition of Done
Shifted automate-backend-deployment component to automate and it can be built using habitat builder.

### :athletic_shoe: How to Build and Test the Change
To build this package

- Cloning this PR changes

- Enter hab studio using `hab studio enter`

- Build respective components using `build components/automate-backend-deployment`

- You will see this on terminal after building the component `/hab/cache/artifacts/meet58-automate-backend-deployment-1.0.28-20210831083109-x86_64-linux.hart' -> '/src/results/meet58-automate-backend-deployment-1.0.28-20210831083109-x86_64-linux.hart`. Where in your case package timestamp and origin name `(E.g meet58 and 20210831083109)`would be different.

- Then install above *.hart file using this command. `hab pkg install /src/results/meet58-automate-backend-deployment-1.0.28-20210831083109-x86_64-linux.hart`

- Then go to `/hab/a2_deploy_workspace`. 

- In that directory you will see Makefile, terraform, certs, inspec, scripts, and test folders. These all meant to be copied from repo to here to run a2ha stuff. (Refere this docs to know more. https://github.com/chef/a2-ha-backend/blob/main/docs/installation.md). 

For this PR, a2ha dir is integrated with automate. For example automate has also inspec dir and a2ha has also inspec dir. So We have to combine both inspes. But during this deployment package building time only a2ha related stuff would get build and pushed in habitat builder. This isolation can be find in component/automate-backend-deployment/habitat/plan.sh file

### :white_check_mark: Checklist

**All PRs** must tick these:


With occasional exceptions, all PRs **from Progress employees** must tick these:


All PRs **from Progress employees** should tick these if appropriate:


*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable
